### PR TITLE
New user levels feature - beginner, intermediate and advanced

### PR DESCRIPTION
**Pull request by oscarfont**
New **user levels feature**. It consists in adding to the users one more configuration parameter for their dailyWord-bot, which is the level of the words they want to be sent. The **possible levels** are: **beginner, intermediate and advanced**.

The user has now **three new commands** available:

- **/mylevels** : Which returns the user's preference list of word levels.
- **/addlevel**: Which lets the user add one level of words to his preference list of word levels.
- **/removelevel**: Which lets the user remove one level of words from his preference list of word levels.

This new **feature** has direct **effects** to the **word bank spreadsheet**. It is needed to be added a **new column** called **level** between word_id and Deutsch which will indicate the level of the word **for each row**. So if this feature is wanted to be implemented there are two things needed to do after the merge to the main project:

1.  **Edit the word bank production spreadsheet**. Add the column level and assign one of the possible levels (beginner, intermediate or advanced) to each of the spreadsheet words.
2. **Ask the users to restart the bot** (call /stop and then /start) so they get assigned a default list of levels (which will be all levels) so then they can play and tweak it around.

### CODE REVIEW

File: daily_word_bot/app.py [hunk 1]
```diff
@@ -25,7 +25,10 @@
     BotCommand("/help", "Opens help section"),
     BotCommand("/start", "Starts sending words"),
     BotCommand("/stop", "Stops sending words"),
-    BotCommand("/blockedwords", "Shows your blocked words")
+    BotCommand("/blockedwords", "Shows your blocked words"),
+    BotCommand("/mylevels", "Shows the level of the words you want to be sent: beginner, intermediate or advanced"),
+    BotCommand("/addlevel", "Adds to your levels a level of the words you wish to receive"),
+    BotCommand("/removelevel", "Removes from your levels a level of the words you wish to receive"),
```
Comment (rooom13) [inline daily_word_bot/app.py:31]: I got an enhancement for this 💡 

/mylevels commands shows in the reply_markup the list of current_levels stored. So that the user can toggle the levels just by clicking on them:

something like:
```
                                                                          -------------
                                                                         | /mylevels   |___
                                                                         ------------- ----
-----------------------------------------------
| Those are your levels:                      |
| - click ➕ for adding or ❌ to removing it |
-----------------------------------------------
| ❌beginner                                  |
-----------------------------------------------
| ❌intermediate                              |
-----------------------------------------------
| ➕advanced                                  |
-----------------------------------------------


```

In this way the user does not need to type the level value but just click on the button.
Including the an emoji would determine which levels are currently active for the user

Comment (rooom13) [reply daily_word_bot/app.py:31]: If done so commands **/addlevel** & **/removelevel** are then not needed which make the conversation flow more user friendly 😄 
 -  you would then include the callback's in `inline_keyboard_callbacks` method

**Review by rooom13 — CHANGES_REQUESTED** [review]

File: daily_word_bot/db.py [hunk 2]
```diff
@@ -11,30 +11,33 @@ def __init__(self, host: str):
         port: int = 6379
         self.r: redis.Redis = redis.Redis(host=host, port=port)
 
-    def save_user(self, message: Message):
+    def save_user(self, message: Message, user_levels: list = []):
```
[the remark this answers is not in the document]
Comment (oscarfont) [reply daily_word_bot/db.py:14]: what an eye! I'll have a look at it

File: daily_word_bot/word_bank.py [hunk 3]
```diff
@@ -46,12 +46,16 @@ def update(self) -> None:
...
 
-    def get_random(self, exclude: list = []) -> dict:
-        """Get a random word excluding the provided ones"""
+    def get_random(self, exclude: list = [], levels: list = []) -> dict:
+        """Get a random word excluding the provided ones and taking into account the user levels"""
         if len(exclude) >= len(self.df.index):
             exclude = []
 
-        df_candidates = self.df.loc[~self.df.index.isin(exclude)]
+        df_candidates = self.df.loc[(~self.df.index.isin(exclude)) & (self.df['level'].isin(levels))]
```
[the remark this answers is not in the document]
Comment (oscarfont) [reply daily_word_bot/word_bank.py:54]: Damn, that's actually true! I'll check it out as well and see what's the best solution

File: daily_word_bot/app.py [hunk 4]
```diff
@@ -177,7 +262,7 @@ def run(self):  # pragma: no cover
             trigger="cron",
             day="*",
             hour="10,18,20",
-            minute="30",
+            minute="8",
```
[the remark this answers is not in the document]
Comment (oscarfont) [reply daily_word_bot/app.py]: Yeah I forgot to change it back hahaha

File: tests/resources/word_bank.csv [hunk 5]
```diff
@@ -1,6 +1,6 @@
...
+word_id;level;Deutsch;Deutscher Ausdruck 1;Deutscher Ausdruck 2;Deutscher Ausdruck 3;Deutscher Ausdruck 4;Spanisch;Spanischer Ausdruck 1;Spanischer Ausdruck 2;Spanischer Ausdruck 3;Spanischer Ausdruck 4
+WID1;intermediate;ab und zu;Vielleicht darf ich ihn mir ab und zu ausleihen;;;;de vez en cuando;Tal vez me lo puedes prestar de vez en cuando.;;;
+WID2;intermediate;abändern;Niemand will die Genfer Konvention abändern, aber wir können die Konvention ergänzen.;;;;modificar;Nadie quiere modificar el Convenio de Ginebra, pero sí que podemos complementarlo.;;;
+WID3;beginner;abdrehen;Wir müssen einen Hahn abdrehen, der ständig offen ist.;;;;cerrar;Es preciso cerrar un grifo permanentemente abierto.;;;
+WID4;advanced;abhalten;Trotzdem, selbst das sollte Dich nicht abhalten.;Ich könnte bis in diese Nacht zurück und sie abhalten, mich wegzugeben.;;;impedir/detener;De todas formas, eso no debería detenerte.;Puedo regresar a aquella noche e impedir que me deje.;;
+WID5;beginner;abholen;Seine Wäsche steht zum Abholen bereit.;Wenn du mich abholst, bin ich deine beste Freundin.;;;recoger/venir a buscar;Su colada está lista para recoger.;Si me vienes a buscar, seré tu mejor amiga.;;
```
[the remark this answers is not in the document]
Comment (oscarfont) [reply tests/resources/word_bank.csv:6]: oh, that's a good point. we could maybe have an "all" level or something like that so is valid for all levels

File: daily_word_bot/app.py (hunk 1 above)
Comment (oscarfont) [reply daily_word_bot/app.py:31]: You are right, it would be much smoother and nicer. I though about the inlines but I though it would be too much effort for a first mvp version of the feature. But I really like the idea, I'll work on it hehe

File: daily_word_bot/db.py [hunk 6]
```diff
@@ -62,6 +65,27 @@ def remove_user_blocked_word(self, message, word_id):
...
         return to_string_list(self.r.smembers(f"blockedWords-{chat_id}"))
 
+    def get_user_levels(self, chat_id) -> typing.List[str]:
+        user_info = self.get_user(chat_id)
+        user_levels = user_info["levels"] if "levels" in user_info else []
+        return user_levels
+
+    def remove_user_level(self, chat_id, level) -> None:
+        user_info = self.get_user(chat_id)
+        levels = user_info["levels"]
```
[the remark this answers is not in the document]
Comment (oscarfont) [reply daily_word_bot/db.py:75]: true! But I thought that implementing this feature in production would mean an update of the users making each one have the column with a default value. What do you think about that?

[the remark this answers is not in the document]
Comment (rooom13) [reply daily_word_bot/db.py:75]: My suggestion would be using `user_info.get("levels")` since this does not raise an Exception when the key is missing.

we could do `user_info.get("levels", [])` to fallback to empty list

File: tests/resources/word_bank.csv (hunk 5 above)
[the remark this answers is not in the document]
Comment (rooom13) [reply tests/resources/word_bank.csv:6]: I would say **null** or empty string should mean "valid for all levels". Then no need to assign a level for all current words (although we'll do it anyways but just in case we left some out)

File: daily_word_bot/app.py (hunk 1 above)
Comment (rooom13) [reply daily_word_bot/app.py:31]: True, if you think it's too much effort I leave it up to you :)

Comment (oscarfont) [reply daily_word_bot/app.py]: I think I solved everything mentioned in the review and I even added the enhancement you proposed. Feel free to check it out and give me some feedback :)

File: daily_word_bot/app.py [hunk 7]
```diff
@@ -102,6 +107,61 @@ def on_get_blockwords_callback(self, update: Update, context: CallbackContext, i
...
+            update.callback_query.edit_message_text(answer.get('msg'), reply_markup=answer.get('reply_markup'))
+        else:
+            update.message.reply_text(answer.get('msg'), reply_markup=answer.get('reply_markup'))
+
+    def on_removelevel_callback(self, update: Update, context: CallbackContext, is_inline_keyboard=False, level_to_remove='') -> None:  # pragma: no cover
+        # get user information from the message
+        message = update.message or update.callback_query.message
+        chat_id = message.chat_id
+
+        self.dao.remove_user_level(chat_id, level_to_remove)
```
Comment (rooom13) [inline daily_word_bot/app.py:132]: from this line you could actually return `on_mylevels_callback` passing the update & context method since the rest of the method's behavior is the same

File: daily_word_bot/app.py [hunk 8]
```diff
@@ -102,6 +107,61 @@ def on_get_blockwords_callback(self, update: Update, context: CallbackContext, i
...
+            update.callback_query.edit_message_text(answer.get('msg'), reply_markup=answer.get('reply_markup'))
+        else:
+            update.message.reply_text(answer.get('msg'), reply_markup=answer.get('reply_markup'))
+
+    def on_addlevel_callback(self, update: Update, context: CallbackContext, is_inline_keyboard=False, level_to_add='') -> None:  # pragma: no cover
+        # get user information from the message
+        message = update.message or update.callback_query.message
+        chat_id = message.chat_id
+
+        self.dao.add_user_level(chat_id, level_to_add)
```
Comment (rooom13) [inline daily_word_bot/app.py:151]: from this line you could actually return `on_mylevels_callback` passing the update & context method since the rest of the method's behavior is the same

File: daily_word_bot/app.py [hunk 9]
```diff
@@ -148,12 +214,17 @@ def send_word(self, context: CallbackContext):  # pragma: no cover
...
-                word_data = self.word_bank.get_random(exclude=exclude)
+                levels = self.dao.get_user_levels(chat_id)
+                word_data = self.word_bank.get_random(exclude=exclude, levels=levels)
+                msg: str = ''
 
-                msg: str = utils.build_word_msg(word_data)
+                if not word_data:
+                    msg = 'Du hast alles gelernt! - ¡Te lo has aprendido todo!'
+                else:
+                    msg = utils.build_word_msg(word_data)
```
Comment (rooom13) [inline daily_word_bot/app.py:224]: `msg =  utils.build_word_msg(word_data) if word_date else 'Du hast alles gelernt! - ¡Te lo has aprendido todo!'`

is a more pythonic way ;)

File: daily_word_bot/db.py [hunk 10]
```diff
@@ -11,30 +12,36 @@ def __init__(self, host: str):
         port: int = 6379
         self.r: redis.Redis = redis.Redis(host=host, port=port)
 
-    def save_user(self, message: Message):
+    def save_user(self, message: Message, user_levels: list = utils.POSSIBLE_USER_LEVELS):
```
Comment (rooom13) [inline daily_word_bot/db.py:15]: let's rather not assign default variables for list arguments., because ...
1. the potential flaw of having objects as default (commented on the previous review)
2. we move business logic to app, so classes have less cross module responsibility. Ideally DAO does not know anything about business logic just manages the db.

File: tests/test_word_bank.py [hunk 11]
```diff
@@ -7,10 +7,14 @@
```
Comment (rooom13) [inline tests/test_word_bank.py:7]: I'd add some tests for different user levels + edge cases.

I'd suggest using `pytest.mark.parametrize`, to save some lines. You can check an example out in _dailyWord-bot/tests/test_utils.py_

**Review by rooom13 — CHANGES_REQUESTED** [review]
💯 delicious solid improvements, check just some other comments to polish it :)

File: daily_word_bot/app.py [hunk 12]
```diff
@@ -102,6 +107,45 @@ def on_get_blockwords_callback(self, update: Update, context: CallbackContext, i
...
+        message = update.message or update.callback_query.message
+        chat_id = message.chat_id
+
+        # look for the levels of the user in the db
+        levels = self.dao.get_user_levels(chat_id)
+
+        # generate answer message and the markup
+        answer = utils.build_levels_answer(levels)
+
+        # answer the user
```
Comment (rooom13) [inline daily_word_bot/app.py:121]: Podràs poner aquí:

msg, reply_markup = answer.get('msg'), answer.get('reply_markup')

File: daily_word_bot/db.py [hunk 13]
```diff
@@ -62,6 +67,27 @@ def remove_user_blocked_word(self, message, word_id):
...
         return to_string_list(self.r.smembers(f"blockedWords-{chat_id}"))
 
+    def get_user_levels(self, chat_id) -> typing.List[str]:
+        user_info = self.get_user(chat_id)
+        user_levels = user_info.get("levels", [])
+        return user_levels
+
+    def remove_user_level(self, chat_id, level) -> None:
+        user_info = self.get_user(chat_id)
+        levels = user_info.get("levels", [])
```
Comment (rooom13) [inline daily_word_bot/db.py:77]: Estas líneas te las puedes ahorrar con el método de arriba get user levels. It applies as well for the method below :)

**Review by rooom13 — CHANGES_REQUESTED** [review]

File: daily_word_bot/utils.py [hunk 14]
```diff
@@ -1,14 +1,60 @@
...
+LEVEL_ASSIGNED: int = 1
+LEVEL_UNASSIGNED: int = 0
+
 
 def highlight(w: str) -> str:
     return f"<b>{w}</b>"
 
 
+def get_levels_tuple_list(user_levels: list) -> list:
+    # variables needed for the method
```
Comment (rooom13) [inline daily_word_bot/utils.py:17]: Discussed in telegram

**Review by rooom13 — CHANGES_REQUESTED** [review]

File: daily_word_bot/db.py (hunk 13 above)
Comment (oscarfont) [reply daily_word_bot/db.py:77]: pero necesito la info de user_info y la funcion solo devuelve los levels del usuario, como mucho podria cambiar user_info.get("levels", []) por get_user_levels pero aun y asi es como si la linea get_user se repitiera

File: daily_word_bot/utils.py [hunk 15]
```diff
@@ -1,14 +1,36 @@
...
 
 
+def build_levels_answer(user_levels: list) -> dict:
+    # build the message and send it back to the user
+    msg = "🛠 Choose the level of the words to be sent.\nClick the empty checkbox ⬜️ to assign or the filled one ✅ to unassign a level. 🛠\n\nThese are your word levels: "
+
+    # create inline keyboard buttons
+    inline_keyboard_buttons = []
+    for level in POSSIBLE_USER_LEVELS:
+        if(level not in user_levels):
```
Comment (rooom13) [inline daily_word_bot/utils.py:21]: parentesis en el condicional no hacen falta, this is python ;)

File: daily_word_bot/utils.py [hunk 16]
```diff
@@ -1,14 +1,36 @@
...
+            level_message = '⬜️ ' + level
```
Comment (rooom13) [inline daily_word_bot/utils.py:22]: mejor anade aqui tb callback_data = ....
(lo mismo para el else)

y entonces mueves el `inline_keyboard_buttons.append` fuera del if else

**Review by rooom13 — APPROVED** [review]
Awesome pull request ❤️ Looking forward seeing this in produciton

**Review by rooom13 — APPROVED** [review]
Solid pr
